### PR TITLE
modernize packaging and add release automation

### PR DESCRIPTION
**Pull request by dhellmann**
This PR updates the packaging files to use pyproject.toml instead of
setup.py and setup.cfg.

It also adds dynamic version computation using setuptools-scm, to derive
the version of the package being built from the tag, so that there is
only one source of truth for the version number.

Finally, it adds a github action that will trigger when a tag is pushed
to build and upload the sdist and wheel files for the new version. There
are some manual steps that need to be completed before that github action
will work, to give github permission to publish to pypi.org.

Related to #269

### CODE REVIEW

**Comment by dhellmann** [comment]
@martinpopel , this is related to #269

**Comment by dhellmann** [comment]
It looks like the linter doesn't like the way I replaced `__version__` in `__main__.py`.

**Comment by martinpopel** [comment]
Great thanks.
I've run the CI and I [see](https://github.com/mjpost/sacrebleu/actions/runs/10080900957/job/27963478472?pr=270#step:9:43) the ruff linter still does not like the `from .version import __version__` line because `F401 .version.__version__ imported but unused` even if `__version__` is now included in `__all__`.
I don't understand why. Or is CI confused by the last forced push?
Should we add `# noqa: F401` to the import line?

**Comment by dhellmann** [comment]
> Great thanks. I've run the CI and I [see](https://github.com/mjpost/sacrebleu/actions/runs/10080900957/job/27963478472?pr=270#step:9:43) the ruff linter still does not like the `from .version import __version__` line because `F401 .version.__version__ imported but unused` even if `__version__` is now included in `__all__`. I don't understand why. Or is CI confused by the last forced push? Should we add `# noqa: F401` to the import line?

I spotted a missing comma when I ran ruff locally, so I've pushed a new version. Let's see if CI likes that better.

**Comment by martinpopel** [comment]
This missing comma is fixed, but we have [another problem](https://github.com/mjpost/sacrebleu/actions/runs/10111815533/job/27964675377?pr=270#step:10:22). I admit I never really understood the imports in sacrebleu.
It is a question whether we still need the `compat.py` module for old style API access (<= 1.4.10).
However, even if we decided to remove it, I would prefer to release at least one version where these methods are issuing a deprecation warning.

**Comment by dhellmann** [comment]
I can fix the circular import problem by importing from the new location.

I don't see a tox environment for running the tests, how do you usually do that locally?

**Comment by dhellmann** [comment]
I worked out how to run the tests, and the latest draft passed a bunch of them locally before I pushed it up for validation in CI.

**Comment by dhellmann** [comment]
I missed some places where automation was using the old approach to building packages.

**Comment by martinpopel** [comment]
The build now [fails with](https://github.com/mjpost/sacrebleu/actions/runs/10117535642/job/28001732454?pr=270#step:12:29):
```
/tmp/build-env-uf1gjo1m/lib/python3.12/site-packages/setuptools_scm/git.py:167: UserWarning: "/home/runner/work/sacrebleu/sacrebleu" is shallow and may cause errors
  warnings.warn(f'"{wd.path}" is shallow and may cause errors')
error: Multiple top-level packages discovered in a flat-layout: ['data', 'sacrebleu'].
```

**Comment by dhellmann** [comment]
> The build now [fails with](https://github.com/mjpost/sacrebleu/actions/runs/10117535642/job/28001732454?pr=270#step:12:29):
> 
> ```
> /tmp/build-env-uf1gjo1m/lib/python3.12/site-packages/setuptools_scm/git.py:167: UserWarning: "/home/runner/work/sacrebleu/sacrebleu" is shallow and may cause errors
>   warnings.warn(f'"{wd.path}" is shallow and may cause errors')
> error: Multiple top-level packages discovered in a flat-layout: ['data', 'sacrebleu'].
> ```

I've been trying to keep this change as small as possible, but there seem to be a few non-standard things about the layout and packaging of the source tree. I could go all in and fix all that, if you're interested? I don't want to overstep, though.

**Comment by dhellmann** [comment]
Let's see if specifying the finder parameters is enough.

**Review by martinpopel — APPROVED** [review]

**Comment by martinpopel** [comment]
I like this, thanks. @mjpost do you agree with merging (as it is mostly you who publishes to PyPI)?

File: sacrebleu/__init__.py
Comment (dhellmann) [inline sacrebleu/__init__.py:1]: I have my editor configured to auto-format using ruff, and that has introduced a lot more change in these files than is truly needed.

File: sacrebleu/__init__.py [hunk 1]
```diff
@@ -14,30 +14,53 @@
...
+    "CHRF",
+    "TER",
+    "corpus_bleu",
+    "raw_corpus_bleu",
+    "sentence_bleu",
+    "corpus_chrf",
+    "sentence_chrf",
+    "corpus_ter",
+    "sentence_ter",
+    "__version__",
```
Comment (dhellmann) [inline sacrebleu/__init__.py:65]: Adding `__version__` here was needed.

File: sacrebleu/__init__.py [hunk 2]
```diff
@@ -14,30 +14,53 @@
...
+from .utils import (
+    SACREBLEU_DIR,
+    download_test_set,
+    get_available_testsets,
+    get_langpairs_for_testset,
+    get_reference_files,
+    get_source_file,
+    smart_open,
+)
+from .version import __version__
```
Comment (dhellmann) [inline sacrebleu/__init__.py:42]: This import replaces the old static definition.

File: sacrebleu/metrics/base.py [hunk 3]
```diff
@@ -8,12 +8,12 @@
 import json
 import logging
 import statistics
-from typing import List, Sequence, Any, Optional, Dict
 from abc import ABCMeta, abstractmethod
+from typing import Any, Dict, List, Optional, Sequence
 
-from .. import __version__
+from ..version import __version__
```
Comment (dhellmann) [inline sacrebleu/metrics/base.py:14]: This import change eliminates a circular dependency between the `sacrebleu` package and this module.

**Comment by mjpost** [comment]
Thanks, @dhellmann! This looks good to me. Don't worry about overstepping, we appreciate the contribution.

I assume I have to upload a secret somewhere for pypi authentication. It wasn't obvious to me what to do, however. Can you comment?

**Comment by mjpost** [comment]
Hmm, I found the link to the docs that you provided. I'll go through the OpenID authorization.

**Comment by mjpost** [comment]
> > The build now [fails with](https://github.com/mjpost/sacrebleu/actions/runs/10117535642/job/28001732454?pr=270#step:12:29):
> > ```
> > /tmp/build-env-uf1gjo1m/lib/python3.12/site-packages/setuptools_scm/git.py:167: UserWarning: "/home/runner/work/sacrebleu/sacrebleu" is shallow and may cause errors
> >   warnings.warn(f'"{wd.path}" is shallow and may cause errors')
> > error: Multiple top-level packages discovered in a flat-layout: ['data', 'sacrebleu'].
> > ```
> 
> I've been trying to keep this change as small as possible, but there seem to be a few non-standard things about the layout and packaging of the source tree. I could go all in and fix all that, if you're interested? I don't want to overstep, though.

Should I merge the current PR? Or did you want to go ahead and fix this as well? We could also consider this as a second PR, if you are interested.

**Comment by dhellmann** [comment]
> > > The build now [fails with](https://github.com/mjpost/sacrebleu/actions/runs/10117535642/job/28001732454?pr=270#step:12:29):
> > > ```
> > > /tmp/build-env-uf1gjo1m/lib/python3.12/site-packages/setuptools_scm/git.py:167: UserWarning: "/home/runner/work/sacrebleu/sacrebleu" is shallow and may cause errors
> > >   warnings.warn(f'"{wd.path}" is shallow and may cause errors')
> > > error: Multiple top-level packages discovered in a flat-layout: ['data', 'sacrebleu'].
> > > ```
> > 
> > 
> > I've been trying to keep this change as small as possible, but there seem to be a few non-standard things about the layout and packaging of the source tree. I could go all in and fix all that, if you're interested? I don't want to overstep, though.
> 
> Should I merge the current PR? Or did you want to go ahead and fix this as well? We could also consider this as a second PR, if you are interested.

Let's leave it as a second PR. I have some deadlines at work that may mean it'll take a while before I can get to it, but if you file an issue and assign it to me I'll come back to it.

**Comment by mjpost** [comment]
Great, thanks, Doug!

**Comment by martinpopel** [comment]
Within this PR, the CI tests work without any errors, but strangely after merging to master, there is [an error](https://github.com/mjpost/sacrebleu/actions/runs/10151812795/job/28071892972#step:6:1):
```
Run python -m build
/opt/hostedtoolcache/Python/3.[11](https://github.com/mjpost/sacrebleu/actions/runs/10151812795/job/28071892972#step:6:12).9/x64/bin/python: No module named build.__main__; 'build' is a package and cannot be directly executed
Error: Process completed with exit code 1.
```
Any ideas how to fix it?

**Comment by dhellmann** [comment]
This line in the log output looks suspicious

```
WARNING: sacrebleu 0.1.dev1+g5af5a3a does not provide the extra 'build'
```

It's probably coming from this line https://github.com/mjpost/sacrebleu/pull/270/files#diff-87c8be2ac3b248aef84cc474af838d7cc461b7f8fb7f5444ac75f4d8fc02e377R32

I used "[build]" there instead of "[dev]" so I think it's not installing the "build" package at all.

**Pull request merged**